elmslie-core/src/main/java/vivid/money/elmslie/core/store/BindingContract.kt

```diff
@@ -0,0 +1,52 @@
+package vivid.money.elmslie.core.store
+
+import io.reactivex.rxjava3.core.Observable
+import io.reactivex.rxjava3.disposables.CompositeDisposable
+
+/**
+ * A contract for passing data between two stores
+ */
+class BindingContract<ParentEvent, ParentEffect, ParentState, ChildEvent, ChildEffect, ChildState>(
+    val parent: Store<ParentEvent, ParentEffect, ParentState>,
+    val child: Store<ChildEvent, ChildEffect, ChildState>,
+) {
+
+    private val disposable = CompositeDisposable()
+    private val contracts = mutableSetOf<() -> Unit>()
+
+    /**
+     * Defines data transformation before passing to another store
+     *
+     * Sample: `parent.states to child using { ChildEvent.OnParentUpdated(it) }`
+     */
+    infix fun <T : Any, Event> Pair<Observable<T>, Store<Event, *, *>>.using(
+        transformation: (T) -> Event
+    ) {
+        contracts += { disposable.add(first.map(transformation).subscribe(second::accept)) }
+    }
+
+    /**
+     * Joins two data sources
+     *
+     * Sample: `parent.states with parent.effects`
+     */
+    infix fun <T : Any, U : Any> Observable<T>.with(
+        other: Observable<U>
+    ): Observable<Pair<T, U>> = Observable.combineLatest(this, other, { f, s -> f to s })
```

Comment:
Seems like pretty dangerous API, since user might be thinking that it's an easy way to retrieve state when mapping effect to event instead of providing necessary data within effect itself but in reality, due to state update, effect mapping could trigger event mapping (and thus accepting) more then once and that might be undesired behaviour

Reply:
Yep, that's a valid problem, but i don't see a clear workaround here. I guess the api could be extended with a number of predefined explicit mappings, restricted by specific cases like this one. But i don't want to limit the api until there's a somewhat complete list of mapping scenarios. What do you think about keeping the api as is and limiting it in future, once the cases will be clear?

Reply:
Yeah, sure